examples/model-viewer/main.cpp

```diff
@@ -0,0 +1,922 @@
+        ComPtr<ISlangBlob> diagnosticsBlob;
+        SlangResult result = device->getSlangSession()->createCompositeComponentType(
+            componentTypes.getBuffer(),
+            componentTypes.getCount(),
+            composedProgram.writeRef(),
+            diagnosticsBlob.writeRef());
+        diagnoseIfNeeded(diagnosticsBlob);
+        SLANG_RETURN_ON_FAIL(result);
+        slangReflection = composedProgram->getLayout();
+
+        // At this point, `composedProgram` represents the shader program
+        // we want to run, and the compute shader there have been checked.
+        // We can create a `gfx::IShaderProgram` object from `composedProgram`
+        // so it may be used by the graphics layer.
+        gfx::IShaderProgram::Desc programDesc = {};
+        programDesc.pipelineType = gfx::PipelineType::Graphics;
+        programDesc.slangProgram = composedProgram.get();
+
+        shaderProgram = device->createProgram(programDesc);
+
+        // Get other shader types that we will use for creating shader objects.
+        perViewShaderType = slangReflection->findTypeByName("PerView");
+        perModelShaderType = slangReflection->findTypeByName("PerModel");
+
+        return SLANG_OK;
+    }
+};
+
+// Our application code has a rudimentary material system,
+// to match the `IMaterial` abstraction used in the shade code.
+//
+struct Material : RefObject
+{
+    // The key feature of a matrial in our application is that
+    // it can provide a shader object that describes it and
+    // its parameters. The contents of the shader object will
+    // be any colors, textures, etc. that the material needs,
+    // while the Slang type that was used to allocate the
+    // block will be an implementation of `IMaterial` that
+    // provides the evaluation logic for the material.
+
+    // Each subclass of `Material` will provide a routine to
+    // create a shader object that stores its shader parameters.
+    virtual IShaderObject* createShaderObject(RendererContext* context) = 0;
+
+    // The shader object for a material will be stashed here
+    // after it is created.
+    ComPtr<IShaderObject> shaderObject;
+};
+
+// For now we have only a single implementation of `Material`,
+// which corresponds to the `SimpleMaterial` type in our shader
+// code.
+//
+struct SimpleMaterial : Material
+{
+    glm::vec3   diffuseColor;
+    glm::vec3   specularColor;
+    float       specularity;
+
+    // Create a shader object that contains the type info and parameter values
+    // that represent an instance of `SimpleMaterial`.
+    IShaderObject* createShaderObject(RendererContext* context) override
+    {
+        auto program = context->slangReflection;
+        auto shaderType = program->findTypeByName("SimpleMaterial");
+        shaderObject = context->device->createShaderObject(shaderType);
+        gfx::ShaderCursor cursor(shaderObject);
+        cursor["diffuseColor"].setData(&diffuseColor, sizeof(diffuseColor));
```

Comment:
Not needed for this PR at all, but it would be great if we can get the syntax `cursor["name"] = value;` to work more broadly. Ideally that operation could be a templated `operator=` in `ShaderCursor` that routes through some user-overridable policy function that the user can then specialize for their own types. This would allow something like `model-viewer` to make use of an adapter that allows setting `glm` matrix/vector types directly into shader cursors.